### ISSUE TITLE
Constraints on ID size

### ISSUE THREAD

**Issue by SylvainCorlay** [ipython/ipykernel#253]
A lot of messages contain strings that are (typically) 32 bytes hexadecimal strings.

As I am working on a native implementation of the protocol, I realize that a lot of optimization is possible in e.g. comms if we know the size of the id in advance. Basically, a comm contains is a string, and fixed-sized stack allocated strings would prevent requiring dynamic allocation when instantiating a comm.

Hence, after some discussions on gitter, a proposition is to require the IDs specified in the jupyter messaging protocol to have a length of *up to* `32 bytes`.

**Comment by SylvainCorlay** [comment]
Note by @minrk:


_There may be some places in Python doing str(uuid.uuid4()) instead of uuid.uuid4().hex, which includes 4 - characters, and is thus 36 bytes.
That should be kept valid. So maybe call it "up to 64 bytes" to give a bunch of headroom, unless there's a reason that's an issue_.

**Comment by SylvainCorlay** [comment]
cc @gouarin @johanmabille

**Comment by JohanMabille** [comment]
As discussed earlier, imposing a *maximal* size seems reasonable. We suggest a maximum number of characters of 55 for the following reasons:
- strings must be null terminated which leads to 56 chars
- since it is a *maximal* size, you need to store the number of char actually used. The common type for that is size_t which on most architectures is 8 bytes long.

This gives a memory footprint of 64 bytes.

**Comment by JohanMabille** [comment]
Choosing for instance 56 or 57 characters would not lead to a memory footprint of 65 or 66 bytes but 72 or 80 bytes (depending on the architecture) due to padding.

**Comment by JohanMabille** [comment]
@minrk Would you accept a PR to the documentation specifying the maximal number of characters in ids?

**Comment by Carreau** [comment]
I think adding on the **current** spec a maximum is fine (as long as it is above what is currently use anywhere), we can revise the spec in a later version to be an **exact** number, I think that would be great. 

Though if we do that, i'd like to have a couple of changes bundle together, instead of having a rolling spec.


But in general +1

**Comment by SylvainCorlay** [comment]
Having a fixed size was the initial idea but in the current spec people can use arbitrary values for ids such as "foobar". So only specifying a maximum length is a less disruptive change which allows us to use stack allocation.

Stack allocated strings make json parsing and comms much much faster. A lot of messages in jupyter contain ids and it would make xeus a lot faster to have that guarantee in the spec.

**Comment by Carreau** [comment]
Sorry, let me be clearer and shorter:
  +1 to add a maximum length to current spec. 
  +1 to make it a fixed length in a further iteration of the spec (like next major revision).

**Comment by minrk** [comment]
Yup, setting a max length of 55 characters seems AOK to me. I don't think there will be any need to specifying a fixed length at any point.
